### PR TITLE
Reuse vertices between subsequent draws, if the same

### PR DESCRIPTION
**Pull request by hrydgard**
Avoids decoding and rebinding vertex data if the current draw uses the same vertex data as the previous draw (but potentially different transform etc).

This helps performance in some games that use "instancing", like GTA and Burnout. With this, vertex cache in VK is pretty much performance neutral in these games, making the case stronger for removing it, and improves performance in GTA with GL by about 6% (measured on PC).

Vertex cache is still a (small) win in some games though - and when there are a lot of repeat, this code now makes the vertex cache path faster too by simply skipping over it on repeats.

~~Looking further back in the frame than just the previous draw and reusing decoded data can probably get even closer or even surpass vertex cache performance (due to simpler management) while pretty much eliminating the risk for bad reuse.~~  Actually, trying this, it doesn't seem all that profitable..

~~Still experimental, needs a lot of testing and also to be implemented for the other backends.~~

### CODE REVIEW

File: GPU/Vulkan/StencilBufferVulkan.cpp
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp:222]: nit: Args are backwards.  Happens to be equivalent.

-[Unknown]

File: GPU/Vulkan/StencilBufferVulkan.cpp
Comment (unknownbrackets) [inline GPU/Vulkan/StencilBufferVulkan.cpp:195]: nit: Backwards here too.

-[Unknown]

File: GPU/GLES/DrawEngineGLES.cpp
Comment (unknownbrackets) [inline GPU/GLES/DrawEngineGLES.cpp:287]: Could we still guard this on some ini setting, even if it's not exposed?  I'd worry that we run into a case where this hash is wrong.  It'd be nice to be able to troubleshoot without a new build.

-[Unknown]

File: GPU/GPUCommon.cpp
Comment (unknownbrackets) [inline GPU/GPUCommon.cpp:573]: Should GLES do this too?

Personally, I'd prefer to clear `prevDcid_` each time we start a new interpret cycle, in case games modify vert data between.  Do the games you're targeting reuse between or only within a single stall?

-[Unknown]

Comment (hrydgard) [reply GPU/GPUCommon.cpp:574]: Yeah, maybe that's a better time to clear prevDcid_. I'm still experimenting.

File: GPU/GLES/DrawEngineGLES.cpp
Comment (hrydgard) [reply GPU/GLES/DrawEngineGLES.cpp:289]: Yeah, might be a good idea.

**Comment by unknownbrackets** [comment]
Hm, it might be interesting to get more stats, but thinking about it the results make sense.  Going back 30 probably wouldn't help, after all.  I'd guess the main times games reuse verts would be:
 * In a row, which this already handles.
 * Far back, i.e. when re-rendering the scene for shading/rear view mirror/etc.
 * When rendering models for shadows, maybe not far back but a decent bit back.

-[Unknown]

**Comment by hrydgard** [comment]
Could of course also keep a hashmap of dcid->drawHistoryIndex, and lookup all the way back. But even a larger window does not seem to be very profitable, so not gonna bother I think. Indeed I think caching vertex data between frames as we've done before is the most profitable, but also the most complex and risky..

**Comment by hrydgard** [comment]
OK, this now works in D3D11 too, and also works when vertex cache is on - and the combination is even faster, heh. Anyway, overall a clear win in some games, a wash in others.

D3D9 will need more refactoring first.

File: GPU/GPUCommon.cpp
Comment (unknownbrackets) [inline GPU/GPUCommon.cpp]: Hm, this is still a heavy sync and may impact timing, though I've not really seen this stat useful too often...

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]
I think this makes sense, I still wonder if we can reset the last draw when returning to the CPU... though it's probably fine.

-[Unknown]

File: GPU/GPUCommon.cpp
Comment (hrydgard) [reply GPU/GPUCommon.cpp]: Oops, I didn't actually mean to remove it here.

Comment (hrydgard) [reply GPU/GPUCommon.cpp]: Well the stat was actually added earlier in this PR, and just got lost in the revert of the history thingy. But still, I'll add it in, I like to have it.

Comment (unknownbrackets) [reply GPU/GPUCommon.cpp]: Oh, oops.  I was thinking I've never noticed it before but forgot it was added in this pull.  But yeah, I think it could be useful for looking at timing.

-[Unknown]

**Comment by hrydgard** [comment]
Kind of in two minds about merging this... The win is very small overall, it's mainly GTA that benefits heh. Most games don't have that many repeats really, so the win is really minor and it does add a little bit of complexity.

**Comment by hrydgard** [comment]
Found it has bad rendering bugs in Burnout Dominator (distorted tree polygons) with the vertex cache on. So might just remove that check for now (so it's either vertex cache, or this).

**Comment by Gamemulatorer** [comment]
How about this now? this really improve performance on opengl backend by 6%?

**Comment by hrydgard** [comment]
Only in a few games like GTA.

Before I merge this, I want to see if I can make it coexist with vertex cache enabled, and make sure it doesn't regress anything noticeably.

**Comment by hrydgard** [comment]
Rebased this, although still not sure I want it in, due to the complexity and edge cases. Bumping to "Future-Prio".

**Comment by hrydgard** [comment]
Not gonna bother with this one, and if I do I'll do it a bit differently.

**Pull request closed without merging**
